presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/CrossJoinWithArrayContainsToInnerJoin.java

```diff
@@ -126,9 +127,9 @@ public Result apply(FilterNode node, Captures captures, Context context)
         List<RowExpression> andConjuncts = extractConjuncts(filterExpression);
         List<RowExpression> remainingConjuncts = andConjuncts.stream().filter(x -> !x.equals(arrayContainsExpression)).collect(toImmutableList());
 
-        VariableReferenceExpression array = (VariableReferenceExpression) ((CallExpression) arrayContainsExpression).getArguments().get(0);
+        RowExpression array = ((CallExpression) arrayContainsExpression).getArguments().get(0);
         VariableReferenceExpression element = (VariableReferenceExpression) ((CallExpression) arrayContainsExpression).getArguments().get(1);
```

Comment:
Is it possible that `element` can also be RowExpression rather than VariableReferenceExpression? I see that in candidate generation, we also assume that element can be a general RowExpression. If `element` cannot be a RowExpression, maybe we can change the candidate generation function to fix the gap.

Reply:
Good point, I added this when I was developing the optimization for NOT CONTAINS but it seems all cases with an expression get pushed down in a projection. I'll make the function more restrictive and only allow variables for the element

Reply:
Actually, for NOT CONTAINS I do hit this case, for example:
```
presto:tpch> explain with t1 as (select * from (values (1, 'JAPAN'), (2, 'invalid_nation')) t(k, nation)) select t1.k, t1.nation from t1 where not contains((select array_agg(name) from nation), t1.nation);
                                                                                                                            Query Plan                                                                                                          >
...
     - FilterProject[PlanNodeId 20,21][filterPredicate = not(contains(array_agg, CAST(field_0 AS varchar(25)))), projectLocality = LOCAL] => [field:integer, field_0:varchar(14)]                                                               >
         - CrossJoin[PlanNodeId 249] => [field:integer, field_0:varchar(14), array_agg:array(varchar(25))]
        ...
 ```

The element in the contains expression here is CAST(field_0 as varchar(25)). I handle this in the transformation and do the projection push down as part of the optimization.
The same query with contains has the cast pushed down before it gets to the optimization:
```
- FilterProject[PlanNodeId 325,324][filterPredicate = contains(array_agg, cast), projectLocality = LOCAL] => [field:integer, field_0:varchar(14)]
```

Reply:
I added some code to abort the CONTAINS optimization if the element is not a variable:
```
        if (!(element instanceof VariableReferenceExpression)) {
            // not currently supported
            return Result.empty();
        }
```
There are tests but I never hit this case so it's mostly a safe guard to not crash if we ever get there.

The same case for NOT CONTAINS is supported and there are tests to check this.

Reply:
Make it a checkstate(element instanceof VariableReferenceExpression) for readability

Reply:
checkstate will raise an error though - I want the behavior to be "don't optimize" and return original plan. Note that this case was not supported previously as the candidate computation function only looked at variables for both the array and the element arguments of CONTAINS

Reply:
Not here - but later where you explicitly just cast it as variable

Reply:
Maybe filter earlier in function `isSupportedArrayContainsFilter` and checkState here.

Reply:
isSupportedArrayContainsFilter is shared between this optimization and the NOT CONTAINS optimization: for NOT CONTAINS we support this case by doing projection push down during optimization. For the CONTAINS case I couldn't find a use case that would trigger this, so I decided not to support it and rather just return early.
I'll add checkstate before casting as @kaikalur suggests

Reply:
Actually it also exists for the CFONTAINS case, for example a similar query like you have above 
```
presto:tpch> explain with t1 as (select * from (values (1, 'JAPAN'), (2, 'invalid_nation')) t(k, nation)) select t1.k, t1.nation from t1 where contains((select array_agg(name) from nation), t1.nation);
                                                                                                                            Query Plan                                                                             >
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------->
 - Output[PlanNodeId 25][k, nation] => [field:integer, field_0:varchar(14)]                                                                                                                                        >
         Estimates: {source: CostBasedSourceInfo, rows: ? (?), cpu: ?, memory: ?, network: ?}                                                                                                                      >
         k := field (1:101)                                                                                                                                                                                        >
         nation := field_0 (1:107)                                                                                                                                                                                 >
     - FilterProject[PlanNodeId 20,21][filterPredicate = contains(array_agg, CAST(field_0 AS varchar(25))), projectLocality = LOCAL] => [field:integer, field_0:varchar(14)]                                       >
             Estimates: {source: CostBasedSourceInfo, rows: ? (?), cpu: ?, memory: ?, network: ?}/{source: CostBasedSourceInfo, rows: ? (?), cpu: ?, memory: ?, network: ?}                                        >
         - CrossJoin[PlanNodeId 249] => [field:integer, field_0:varchar(14), array_agg:array(varchar(25))]                                                                                                         >
                 Estimates: {source: CostBasedSourceInfo, rows: 2 (120B), cpu: ?, memory: ?, network: ?}                                                                                                           >
                 Distribution: REPLICATED                                                                                                                                                                          >
             - LocalExchange[PlanNodeId 561][ROUND_ROBIN] () => [field:integer, field_0:varchar(14)]                                                                                                               >
                     Estimates: {source: CostBasedSourceInfo, rows: 2 (120B), cpu: 120.00, memory: 0.00, network: 0.00}                                                                                            >
                 - Values[PlanNodeId 0] => [field:integer, field_0:varchar(14)]                                                                                                                                    >
                         Estimates: {source: CostBasedSourceInfo, rows: 2 (120B), cpu: 0.00, memory: 0.00, network: 0.00}                                                                                          >
                         (INTEGER'1', VARCHAR'JAPAN')                                                                                                                                                              >
                         (INTEGER'2', VARCHAR'invalid_nation')                                                                                                                                                     >
             - EnforceSingleRow[PlanNodeId 17] => [array_agg:array(varchar(25))]                                                                                                                                   >
                     Estimates: {source: CostBasedSourceInfo, rows: 1 (60B), cpu: ?, memory: ?, network: ?}                                                                                                        >
                 - Aggregate(FINAL)[PlanNodeId 12] => [array_agg:array(varchar(25))]                                                                                                                               >
                         array_agg := "presto.default.array_agg"((array_agg_29)) (1:148)                                                                                                                           >
                     - LocalExchange[PlanNodeId 578][SINGLE] () => [array_agg_29:array(varchar(25))]                                                                                                               >
                         - RemoteStreamingExchange[PlanNodeId 584][GATHER] => [array_agg_29:array(varchar(25))]                                                                                                    >
                             - Aggregate(PARTIAL)[PlanNodeId 582] => [array_agg_29:array(varchar(25))]                                                                                                             >
                                     array_agg_29 := "presto.default.array_agg"((name)) (1:148)                                                                                                                    >
                                 - TableScan[PlanNodeId 9][TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=nation, analyzePartitionValues=Optional.empty}', layout='Op>
                                         Estimates: {source: CostBasedSourceInfo, rows: 25 (1.46kB), cpu: 302.00, memory: 0.00, network: 0.00}                                                                     >
                                         LAYOUT: tpch.nation{}                                                                                                                                                     >
                                         name := name:varchar(25):1:REGULAR (1:169)                                                                                                                                >
                                                                                                                                                                                                                   >
(1 row)
```

Reply:
That's weird, I don't get that, in my case the cast is pushed down below the join:
```

presto:tpch> explain with t1 as (select * from (values (1, 'JAPAN'), (2, 'invalid_nation')) t(k, nation)) select t1.k, t1.nation from t1 where contains((select array_agg(name) from nation), t1.nation);
                                                                                                                            Query Plan
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[PlanNodeId 25][k, nation] => [field:integer, field_0:varchar(14)]
         Estimates: {source: CostBasedSourceInfo, rows: ? (?), cpu: ?, memory: ?, network: ?}
         k := field (1:101)
         nation := field_0 (1:107)
     - FilterProject[PlanNodeId 325,324][filterPredicate = contains(array_agg, cast), projectLocality = LOCAL] => [field:integer, field_0:varchar(14)]
             Estimates: {source: CostBasedSourceInfo, rows: ? (?), cpu: ?, memory: ?, network: ?}/{source: CostBasedSourceInfo, rows: ? (?), cpu: ?, memory: ?, network: ?}
         - CrossJoin[PlanNodeId 326] => [field:integer, field_0:varchar(14), cast:varchar(25), array_agg:array(varchar(25))]
                 Estimates: {source: CostBasedSourceInfo, rows: 2 (120B), cpu: ?, memory: ?, network: ?}
                 Distribution: REPLICATED
             - Project[PlanNodeId 323][projectLocality = LOCAL] => [field:integer, field_0:varchar(14), cast:varchar(25)]
                     Estimates: {source: CostBasedSourceInfo, rows: 2 (120B), cpu: 350.00, memory: 0.00, network: 0.00}
                     cast := CAST(field_0 AS varchar(25)) (1:22)
                 - LocalExchange[PlanNodeId 621][ROUND_ROBIN] () => [field:integer, field_0:varchar(14)]
                         Estimates: {source: CostBasedSourceInfo, rows: 2 (120B), cpu: 120.00, memory: 0.00, network: 0.00}
                     - Values[PlanNodeId 0] => [field:integer, field_0:varchar(14)]
                             Estimates: {source: CostBasedSourceInfo, rows: 2 (120B), cpu: 0.00, memory: 0.00, network: 0.00}
                             (INTEGER'1', VARCHAR'JAPAN')
                             (INTEGER'2', VARCHAR'invalid_nation')
             - EnforceSingleRow[PlanNodeId 17] => [array_agg:array(varchar(25))]
                     Estimates: {source: CostBasedSourceInfo, rows: 1 (60B), cpu: ?, memory: ?, network: ?}
                 - Aggregate(FINAL)[PlanNodeId 12] => [array_agg:array(varchar(25))]
                         array_agg := "presto.default.array_agg"((array_agg_29)) (1:148)
                     - LocalExchange[PlanNodeId 639][SINGLE] () => [array_agg_29:array(varchar(25))]
                         - RemoteStreamingExchange[PlanNodeId 645][GATHER] => [array_agg_29:array(varchar(25))]
                             - Aggregate(PARTIAL)[PlanNodeId 643] => [array_agg_29:array(varchar(25))]
                                     array_agg_29 := "presto.default.array_agg"((name)) (1:148)
                                 - TableScan[PlanNodeId 9][TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=nation, analyzePartitionValues=Optional.empty}', layout='Optional[tpch.nation{}]'}] => [name:varchar(25)]
                                         Estimates: {source: CostBasedSourceInfo, rows: 25 (1.46kB), cpu: 302.00, memory: 0.00, network: 0.00}
                                         LAYOUT: tpch.nation{}
                                         name := name:varchar(25):1:REGULAR (1:169)

(1 row)
```

Reply:
Oh, now I get all the context back.
The two existing cross join to inner join optimizations are used together with the optimization `PushDownFilterExpressionEvaluationThroughCrossJoin`. This `PushDownFilterExpressionEvaluationThroughCrossJoin` will create projections for the valid contains/or expression, and project the row expressions to VariableReferenceExpression. And that's why in existing `CrossJoinWithArrayContainsToInnerJoin` optimization, it always assume VariableReferenceExpression.

My test is in master branch, and your run is with change from this PR. As the `PushDownFilterExpressionEvaluationThroughCrossJoin` also changed in this PR, hence lead to different output here.

Let's discuss offline on what to do here, as it seems to need more thinking here.

Reply:
I see, getCandidateArrayContainsExpression is now more permissive and allows for expressions, so those get picked up and pushed down in CrossJoinWithArrayContainsToInnerJoin. 
I think the old PushDownFIlter... had a bug: when looking whether there are candidate arrayContains expressions it only checked for variables, but the actual implementation is looking for expressions as well: addCandidateExpression is checking whether the used variables in the expression are present in the corresponding output columns, but with the old implementation of getCandidateArrayContainsExpression I don't think we would have selected such expressions as candidates at all.
Since we have tests to cover this case I think we should be fine

Reply:
The `CrossJoinWithArrayContainsToInnerJoin` relies on the `PushDownFilterExpressionEvaluationThroughCrossJoin` to map expressions to variables, and only consider the case where the input of contains function is variables.
I think the new not array contains optimizer should either follow the same pattern, i.e. rely on the `PushDownFilterExpressionEvaluationThroughCrossJoin` to do pushdown as existing ones, or act as an independent optimizer without touching the existing ones. The current PR leave the three cross join to inner join optimizers mixed together, but actually two of them rely on `PushDownFilterExpressionEvaluationThroughCrossJoin` while one does it by itself.

Reply:
> I think the old PushDownFIlter... had a bug: when looking whether there are candidate arrayContains expressions it only checked for variables, but the actual implementation is looking for expressions as well: addCandidateExpression is checking whether the used variables in the expression are present in the corresponding output columns, but with the old implementation of getCandidateArrayContainsExpression I don't think we would have selected such expressions as candidates at all.
Since we have tests to cover this case I think we should be fine


Would you mind describe it in more detail (or discuss it offline if it's hard to describe in comment here)? I get lost on this part.